File: query/src/main/scala/com/lightbend/modelServer/query/ModelStateQuery.scala
/*
 * Copyright (C) 2017  Lightbend
 *
 * This file is part of flink-ModelServing
 *
 * flink-ModelServing is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.lightbend.modelServer.query

import com.lightbend.modelServer.model.ModelToServeStats
import com.lightbend.modelServer.model.speculative.SpeculativeExecutionStats
import org.apache.flink.api.common.state.ValueStateDescriptor
import org.apache.flink.api.common.{ExecutionConfig, JobID}
import org.apache.flink.api.common.typeinfo.BasicTypeInfo
import org.apache.flink.api.scala.createTypeInformation
import org.apache.flink.queryablestate.client.QueryableStateClient
import org.joda.time.DateTime

import scala.collection.mutable.ListBuffer

/**
  * Created by boris on 5/12/17.
  * see https://ci.apache.org/projects/flink/flink-docs-release-1.3/dev/stream/queryable_state.html
  */
object ModelStateQuery {

  val timeInterval = 1000 * 20        // 20 sec

  def main(args: Array[String]) {

    val jobId = JobID.fromHexString("ee5332de3f8e0fe9320b69d08b1a1b78")
    val types = Array("cpu")

    val client = new QueryableStateClient("127.0.0.1", 9069)

    // State descriptor for individual mpdels
    val currentModelsDesc = new ValueStateDescriptor[ListBuffer[String]](
      "modelList", // state name
      createTypeInformation[ListBuffer[String]].createSerializer(new ExecutionConfig)  // type Serializer
    )

    // State descriptor for voting model.
    val speculativeModelDesc = new ValueStateDescriptor[SpeculativeExecutionStats](
      "currentSpeculativeModel",   // state name
      createTypeInformation[SpeculativeExecutionStats].createSerializer(new ExecutionConfig)  // type Serializer
    )

    // State descriptor for state of individual model.
    val modelStateDesc = new ValueStateDescriptor[ModelToServeStats](
      "currentModelState",                  // state name
      createTypeInformation[ModelToServeStats].createSerializer(new ExecutionConfig)  // type Serializer
    )

    val keyType = BasicTypeInfo.STRING_TYPE_INFO

    while(true) {
      for (key <- types) {
        try {
          // Get model list
          val modelsListFuture = client.getKvState(jobId, "currentModels", key, keyType, currentModelsDesc)
          val models = modelsListFuture.join().value()
          println(s"Currrent models for type $key is $models")
          // Individual models
          models.foreach(model => {
            val modelStateFuture = client.getKvState(jobId, "currentModelState", model, keyType, modelStateDesc)
            val modelState = modelStateFuture.join().value()
            println(s"Model ${modelState.description}, deployed at ${new DateTime(modelState.since).toString("yyyy/MM/dd HH:MM:SS")}}. " +
              s"Average execution ${modelState.duration/modelState.usage}, min execution ${modelState.min}, max execution ${modelState.max}")
          })
          // Speculative model
          val speculativeStateFuture = client.getKvState(jobId, "speculativeState", key, keyType, speculativeModelDesc)
          val speculativeState = speculativeStateFuture.join().value()
          println(s"Speculative Model for type $key is" +
            s"Deployed at ${new DateTime(speculativeState.since).toString("yyyy/MM/dd HH:MM:SS")}} with decider ${speculativeState.decider} " +
            s"and timeout ${speculativeState.tmout}. Average execution ${speculativeState.duration/speculativeState.usage}, min execution ${speculativeState.min}, max execution ${speculativeState.max}")
        }
        catch {case e: Exception => e.printStackTrace()}
      }
      Thread.sleep(timeInterval)
    }
  }
}

File: query/src/main/java/com/lightbend/java/modelServer/query/ModelStateQuery.java
package com.lightbend.java.modelServer.query;

import com.lightbend.java.modelServer.model.ModelToServeStats;
import com.lightbend.java.modelServer.model.speculative.SpeculativeExecutionStats;
import org.apache.flink.api.common.ExecutionConfig;
import org.apache.flink.api.common.JobID;
import org.apache.flink.api.common.state.ValueState;
import org.apache.flink.api.common.state.ValueStateDescriptor;
import org.apache.flink.api.common.typeinfo.BasicTypeInfo;
import org.apache.flink.api.common.typeinfo.TypeHint;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.queryablestate.client.QueryableStateClient;
import org.joda.time.DateTime;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public class ModelStateQuery {

    private static final int timeInterval = 1000 * 20;        // 20 sec

    public static void main(String[] args) throws Exception {

        JobID jobId = JobID.fromHexString("93e731a1da73551656c459b58a264e62");
        List<String> types = Arrays.asList("cpu");

        QueryableStateClient client = new QueryableStateClient("127.0.0.1", 9069);

        // State descriptor for individual mpdels
        ValueStateDescriptor<List<String>> currentModelsDesc = new ValueStateDescriptor<List<String>>(
                "modelList", // state name
                TypeInformation.of(new TypeHint<List<String>>() {}).createSerializer(new ExecutionConfig())  // type Serializer
        );

        // State descriptor for voting model.
        ValueStateDescriptor<SpeculativeExecutionStats> speculativeModelDesc = new ValueStateDescriptor<>(
                "currentSpeculativeModel",   // state name
                TypeInformation.of(SpeculativeExecutionStats.class).createSerializer(new ExecutionConfig())  // type Serializer
        );

        // State descriptor for state of individual model.
        ValueStateDescriptor<ModelToServeStats> modelStateDesc = new ValueStateDescriptor<>(
                "currentModelState",                  // state name
                TypeInformation.of(ModelToServeStats.class).createSerializer(new ExecutionConfig()) // type serializer
        );

        BasicTypeInfo keyType = BasicTypeInfo.STRING_TYPE_INFO;

        while(true) {
            for (String key : types) {

                // Get model list
                CompletableFuture<ValueState<List<String>>> modelsListFuture = client.getKvState(jobId, "currentModels", key, keyType, currentModelsDesc);
                modelsListFuture.thenAccept(response -> {
                    try{
                        List<String> models = response.value();
                        System.out.println("Currrent models for type " + key + " are " + models);
                    }catch (Throwable e) {
                        e.printStackTrace();
                    }
                });
                List<String> models = modelsListFuture.get().value();

                // Individual models
                for (String model : models){
                    CompletableFuture<ValueState<ModelToServeStats>> modelStateFuture = client.getKvState(jobId, "currentModelState", model, keyType, modelStateDesc);
                    modelStateFuture.thenAccept(response -> {
                        try{
                            ModelToServeStats stats = response.value();
                            System.out.println("Model " + stats.getName() + " deployed at " + new DateTime(stats.getSince()).toString("yyyy/MM/dd HH:MM:SS") +
                                    ".  Average execution " + stats.getDuration()/stats.getDuration() + " min execution " + stats.getMin() + " max execution " + stats.getMax());
                        }catch (Throwable e) {
                            e.printStackTrace();
                        }
                    });
                }

                // Speculative model
                CompletableFuture<ValueState<SpeculativeExecutionStats>> speculativeStateFuture = client.getKvState(jobId, "speculativeState", key, keyType, speculativeModelDesc);
                speculativeStateFuture.thenAccept(response -> {
                    try {
                        SpeculativeExecutionStats stats = response.value();
                        System.out.println("Speculative Model for type " + key + " is " + stats.getName() + " deployed at " +
                                new DateTime(stats.getSince()).toString("yyyy/MM/dd HH:MM:SS") + " decider " + stats.getDecider() + " timeout " + stats.getTmout() +
                                ".  Average execution " + stats.getDuration()/stats.getDuration() + " min execution " + stats.getMin() + " max execution " + stats.getMax());
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                });
            }
            Thread.sleep(timeInterval);
        }
    }
}